README.md
# ade_flutterwave_working_version

A simple Flutterwave payment that works.

## Youtube Preview

[![IMAGE ALT TEXT HERE](https://img.youtube.com/vi/uOvP3rwnVwQ/0.jpg)](https://www.youtube.com/watch?v=uOvP3rwnVwQ)

[![Pub](https://img.shields.io/pub/v/flutter_slidable.svg)](https://pub.dev/packages/ade_flutterwave_working_version)

![Overview](https://github.com/adeleyeayodeji/ade_flutterwave_working_version/blob/main/screenshot2.png)

![Overview](https://github.com/adeleyeayodeji/ade_flutterwave_working_version/blob/main/screenshot1.png)

## Features

- Webview interface
- Simple to use.
- No stress added.
- Many more . . .

## Getting started

In the `pubspec.yaml` of your flutter project, add the following dependency:

```yaml
dependencies:
  ...
  ade_flutterwave_working_version: ^0.0.2+2
```

In your library add the following import:

```dart
import 'package:ade_flutterwave_working_version/core/ade_flutterwave.dart';
```

For help getting started with Flutter, view the online [documentation](https://flutter.io/).

### Constructors

You can create a `AdeFlutterWavePay`:

```dart
  //check if the fields are empty
 if (_amountController.text.isEmpty ||
        _emailController.text.isEmpty ||
        _phoneController.text.isEmpty ||
        _fullNameController.text.isEmpty) {
      ScaffoldMessenger.of(context).showSnackBar(
        const SnackBar(
          content: Text('Please fill all the fields'),
        ),
      );
      return;
    }
    //data
    var data = {
      'amount': _amountController.text,
      'email': _emailController.text,
      'phone': _phoneController.text,
      'name': _fullNameController.text,
      'payment_options': 'card, banktransfer, ussd',
      'title': 'Flutterwave payment',
      'currency': "NGN",
      'tx_ref':
          "AdeFlutterwave-${DateTime.now().millisecondsSinceEpoch}",
      'icon':
          "https://www.aqskill.com/wp-content/uploads/2020/05/logo-pde.png",
      'public_key':
          "FLWPUBK_TEST-e0787ab2e5b0b6fcb3d32ce465ad44d0-X",
      'sk_key':
          'FLWSECK_TEST-af1af523da3f141f894a26be4b071230-X'
    };

    Navigator.push(
      context,
      MaterialPageRoute(
        builder: (context) => AdeFlutterWavePay(data),
      ),
    ).then((response) {
      //response is the response from the payment
      // print(response);
  });
```

### Full Example

```Dart
import 'package:flutter/material.dart';
import 'package:ade_flutterwave_working_version/core/ade_flutterwave.dart';

class AdeFlutterExample extends StatefulWidget {
  const AdeFlutterExample({Key? key}) : super(key: key);

  @override
  State<AdeFlutterExample> createState() => _AdeFlutterExampleState();
}

class _AdeFlutterExampleState extends State<AdeFlutterExample> {
  //controllers
  final TextEditingController _amountController = TextEditingController();
  final TextEditingController _emailController = TextEditingController();
  final TextEditingController _phoneController = TextEditingController();
  final TextEditingController _fullNameController = TextEditingController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: const Text('Ade Flutterwave payment'),
          backgroundColor: Colors.orange,
          foregroundColor: Colors.white,
        ),
        body: Padding(
          padding: const EdgeInsets.all(15.0),
          child: Form(
            child: Column(
              children: [
                //title
                const Text(
                  'Flutterwave payment',
                  style: TextStyle(
                    fontSize: 20,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                TextFormField(
                  controller: _fullNameController,
                  decoration: const InputDecoration(
                    labelText: 'Name',
                  ),
                ),
                TextFormField(
                  controller: _emailController,
                  decoration: const InputDecoration(
                    labelText: 'Email',
                  ),
                ),
                TextFormField(
                  controller: _phoneController,
                  decoration: const InputDecoration(
                    labelText: 'Phone',
                  ),
                ),
                TextFormField(
                  controller: _amountController,
                  decoration: const InputDecoration(
                    labelText: 'Amount',
                  ),
                ),
                ElevatedButton(
                  onPressed: () {
                    //check if the fields are empty
                    if (_amountController.text.isEmpty ||
                        _emailController.text.isEmpty ||
                        _phoneController.text.isEmpty ||
                        _fullNameController.text.isEmpty) {
                      ScaffoldMessenger.of(context).showSnackBar(
                        const SnackBar(
                          content: Text('Please fill all the fields'),
                        ),
                      );
                      return;
                    }
                    //data
                    var data = {
                      'amount': _amountController.text,
                      'email': _emailController.text,
                      'phone': _phoneController.text,
                      'name': _fullNameController.text,
                      'payment_options': 'card, banktransfer, ussd',
                      'title': 'Flutterwave payment',
                      'currency': "NGN",
                      'tx_ref':
                          "AdeFlutterwave-${DateTime.now().millisecondsSinceEpoch}",
                      'icon':
                          "https://www.aqskill.com/wp-content/uploads/2020/05/logo-pde.png",
                      'public_key':
                          "FLWPUBK_TEST-e0787ab2e5b0b6fcb3d32ce465ad44d0-X",
                      'sk_key':
                          'FLWSECK_TEST-af1af523da3f141f894a26be4b071230-X'
                    };

                    Navigator.push(
                      context,
                      MaterialPageRoute(
                        builder: (context) => AdeFlutterWavePay(data),
                      ),
                    ).then((response) {
                      //response is the response from the payment
                      // print(response);
                    });
                  },
                  child: const Text('Pay'),
                ),
              ],
            ),
          ),
        ));
  }
}

```


CHANGELOG.md
## 0.0.2+2

- Fix bugs in the package

## 0.0.2+1

- Added payment options

## 0.0.1+4

- Moved html from assets to string format

## 0.0.1+3

- Updated Readme

## 0.0.1+2

- Updated Readme

## 0.0.1+1

- Initial Open Source release.


test/widget_test.dart
// This is a basic Flutter widget test.
//
// To perform an interaction with a widget in your test, use the WidgetTester
// utility in the flutter_test package. For example, you can send tap and scroll
// gestures. You can also use WidgetTester to find child widgets in the widget
// tree, read text, and verify that the values of widget properties are correct.

import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:ade_flutterwave_working_version/main.dart';

void main() {}


linux/my_application.cc
#include "my_application.h"

#include <flutter_linux/flutter_linux.h>
#ifdef GDK_WINDOWING_X11
#include <gdk/gdkx.h>
#endif

#include "flutter/generated_plugin_registrant.h"

struct _MyApplication {
  GtkApplication parent_instance;
  char** dart_entrypoint_arguments;
};

G_DEFINE_TYPE(MyApplication, my_application, GTK_TYPE_APPLICATION)

// Implements GApplication::activate.
static void my_application_activate(GApplication* application) {
  MyApplication* self = MY_APPLICATION(application);
  GtkWindow* window =
      GTK_WINDOW(gtk_application_window_new(GTK_APPLICATION(application)));

  // Use a header bar when running in GNOME as this is the common style used
  // by applications and is the setup most users will be using (e.g. Ubuntu
  // desktop).
  // If running on X and not using GNOME then just use a traditional title bar
  // in case the window manager does more exotic layout, e.g. tiling.
  // If running on Wayland assume the header bar will work (may need changing
  // if future cases occur).
  gboolean use_header_bar = TRUE;
#ifdef GDK_WINDOWING_X11
  GdkScreen* screen = gtk_window_get_screen(window);
  if (GDK_IS_X11_SCREEN(screen)) {
    const gchar* wm_name = gdk_x11_screen_get_window_manager_name(screen);
    if (g_strcmp0(wm_name, "GNOME Shell") != 0) {
      use_header_bar = FALSE;
    }
  }
#endif
  if (use_header_bar) {
    GtkHeaderBar* header_bar = GTK_HEADER_BAR(gtk_header_bar_new());
    gtk_widget_show(GTK_WIDGET(header_bar));
    gtk_header_bar_set_title(header_bar, "flutterwave_working_version");
    gtk_header_bar_set_show_close_button(header_bar, TRUE);
    gtk_window_set_titlebar(window, GTK_WIDGET(header_bar));
  } else {
    gtk_window_set_title(window, "flutterwave_working_version");
  }

  gtk_window_set_default_size(window, 1280, 720);
  gtk_widget_show(GTK_WIDGET(window));

  g_autoptr(FlDartProject) project = fl_dart_project_new();
  fl_dart_project_set_dart_entrypoint_arguments(project, self->dart_entrypoint_arguments);

  FlView* view = fl_view_new(project);
  gtk_widget_show(GTK_WIDGET(view));
  gtk_container_add(GTK_CONTAINER(window), GTK_WIDGET(view));

  fl_register_plugins(FL_PLUGIN_REGISTRY(view));

  gtk_widget_grab_focus(GTK_WIDGET(view));
}

// Implements GApplication::local_command_line.
static gboolean my_application_local_command_line(GApplication* application, gchar*** arguments, int* exit_status) {
  MyApplication* self = MY_APPLICATION(application);
  // Strip out the first argument as it is the binary name.
  self->dart_entrypoint_arguments = g_strdupv(*arguments + 1);

  g_autoptr(GError) error = nullptr;
  if (!g_application_register(application, nullptr, &error)) {
     g_warning("Failed to register: %s", error->message);
     *exit_status = 1;
     return TRUE;
  }

  g_application_activate(application);
  *exit_status = 0;

  return TRUE;
}

// Implements GObject::dispose.
static void my_application_dispose(GObject* object) {
  MyApplication* self = MY_APPLICATION(object);
  g_clear_pointer(&self->dart_entrypoint_arguments, g_strfreev);
  G_OBJECT_CLASS(my_application_parent_class)->dispose(object);
}

static void my_application_class_init(MyApplicationClass* klass) {
  G_APPLICATION_CLASS(klass)->activate = my_application_activate;
  G_APPLICATION_CLASS(klass)->local_command_line = my_application_local_command_line;
  G_OBJECT_CLASS(klass)->dispose = my_application_dispose;
}

static void my_application_init(MyApplication* self) {}

MyApplication* my_application_new() {
  return MY_APPLICATION(g_object_new(my_application_get_type(),
                                     "application-id", APPLICATION_ID,
                                     "flags", G_APPLICATION_NON_UNIQUE,
                                     nullptr));
}


windows/runner/main.cpp
#include <flutter/dart_project.h>
#include <flutter/flutter_view_controller.h>
#include <windows.h>

#include "flutter_window.h"
#include "utils.h"

int APIENTRY wWinMain(_In_ HINSTANCE instance, _In_opt_ HINSTANCE prev,
                      _In_ wchar_t *command_line, _In_ int show_command) {
  // Attach to console when present (e.g., 'flutter run') or create a
  // new console when running with a debugger.
  if (!::AttachConsole(ATTACH_PARENT_PROCESS) && ::IsDebuggerPresent()) {
    CreateAndAttachConsole();
  }

  // Initialize COM, so that it is available for use in the library and/or
  // plugins.
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);

  flutter::DartProject project(L"data");

  std::vector<std::string> command_line_arguments =
      GetCommandLineArguments();

  project.set_dart_entrypoint_arguments(std::move(command_line_arguments));

  FlutterWindow window(project);
  Win32Window::Point origin(10, 10);
  Win32Window::Size size(1280, 720);
  if (!window.CreateAndShow(L"flutterwave_working_version", origin, size)) {
    return EXIT_FAILURE;
  }
  window.SetQuitOnClose(true);

  ::MSG msg;
  while (::GetMessage(&msg, nullptr, 0, 0)) {
    ::TranslateMessage(&msg);
    ::DispatchMessage(&msg);
  }

  ::CoUninitialize();
  return EXIT_SUCCESS;
}


lib/core/ade_flutterwave.dart
// ignore_for_file: prefer_collection_literals, avoid_print, use_build_context_synchronously, library_private_types_in_public_api, no_logic_in_create_state, prefer_typing_uninitialized_variables, avoid_init_to_null

import 'dart:async';
import 'dart:convert';
import 'package:flutter/material.dart';
import 'package:loading_overlay/loading_overlay.dart';
import 'package:http/http.dart' as http;
import 'package:webview_flutter/webview_flutter.dart';

class AdeFlutterWavePay extends StatefulWidget {
  final data;
  const AdeFlutterWavePay(this.data, {super.key});

  @override
  _AdeFlutterWavePayState createState() => _AdeFlutterWavePayState(data);
}

class _AdeFlutterWavePayState extends State<AdeFlutterWavePay> {
  _AdeFlutterWavePayState(this.data);
  String filepath = "files/flutterwave.html";
  late WebViewController _webViewController;
  dynamic data;
  bool isGeneratingCode = true;
  bool cango = false;
  dynamic result = null;

  loadfunction() {
    Timer(const Duration(seconds: 5), () {
      _webViewController.runJavascriptReturningResult(
          'makePayment("${data["tx_ref"]}", "${data["amount"]}", "${data["email"]}", "${data["title"]}", "${data["name"]}", "${data["currency"]}", "${data["icon"]}", "${data["public_key"]}", "${data["phone"]}", "${data["payment_options"]}")');
      setState(() {
        isGeneratingCode = false;
      });
    });
  }

  @override
  void initState() {
    super.initState();
  }

  @override
  Widget build(BuildContext context) {
    return LoadingOverlay(
      isLoading: isGeneratingCode,
      opacity: 1,
      color: Colors.black54,
      child: Scaffold(
        floatingActionButton: FloatingActionButton(
          child: const Icon(Icons.refresh_sharp),
          onPressed: () {
            _webViewController.runJavascriptReturningResult(
                'makePayment("${data["tx_ref"]}", "${data["amount"]}", "${data["email"]}", "${data["title"]}", "${data["name"]}", "${data["currency"]}", "${data["icon"]}", "${data["public_key"]}", "${data["phone"]}", "${data["payment_options"]}")');
          },
        ),
        appBar: AppBar(
          title: Text(data["title"]),
          backgroundColor: Colors.orange,
          foregroundColor: Colors.white,
          leading: IconButton(
            icon: const Icon(Icons.arrow_back),
            onPressed: () async {
              try {
                Future.delayed(const Duration(seconds: 2), () {
                  // WidgetsBinding.instance.addPostFrameCallback((_) {
                  Navigator.of(context).pop();
                  // });
                });
              } catch (e) {
                ScaffoldMessenger.of(context).showSnackBar(const SnackBar(
                  content: Text(
                    "Something went wrong!",
                    style: TextStyle(color: Colors.white),
                  ),
                  backgroundColor: Colors.red,
                  duration: Duration(seconds: 5),
                ));
              }
            },
          ),
        ),
        body: WebView(
          initialUrl: '',
          javascriptMode: JavascriptMode.unrestricted,
          javascriptChannels: Set.from([
            JavascriptChannel(
                name: "messageHandler",
                onMessageReceived: (message) async {
                  var d = message.message;
                  _webViewController.goBack();
                  _webViewController.clearCache();
                  var response = jsonDecode(d);
                  if (response["status"] != "cancelled") {
                    ScaffoldMessenger.of(context).showSnackBar(const SnackBar(
                      content: Text("Verifying transactions, please wait.."),
                      duration: Duration(seconds: 5),
                    ));
                    // print(response);
                    setState(() {
                      isGeneratingCode = true;
                    });
                    try {
                      var response2 = await http.get(
                          Uri.parse(
                              "https://api.flutterwave.com/v3/transactions/${response["transaction_id"]}/verify"),
                          headers: {
                            "Authorization": "Bearer ${data["sk_key"]}",
                            "Content-Type": "application/json",
                            "Accept": "application/json"
                          });
                      var d2 = jsonDecode(response2.body);
                      _webViewController.goBack();
                      _webViewController.clearCache();
                      // print(d2);
                      if (d2["status"] == "success") {
                        ScaffoldMessenger.of(context)
                            .showSnackBar(const SnackBar(
                          content: Text(
                            "Transaction verified!, redirecting...",
                            style: TextStyle(color: Colors.white),
                          ),
                          backgroundColor: Colors.green,
                          duration: Duration(seconds: 10),
                        ));
                        setState(() {
                          result = d2;
                        });
                        Future.delayed(const Duration(seconds: 10), () {
                          if (mounted) {
                            setState(() {
                              isGeneratingCode = false;
                            });
                            WidgetsBinding.instance.addPostFrameCallback((_) {
                              //Moving forward
                              Navigator.pop(context, d2);
                            });
                          }
                        });
                      } else {
                        ScaffoldMessenger.of(context)
                            .showSnackBar(const SnackBar(
                          content: Text(
                            "Transaction failed!",
                            style: TextStyle(color: Colors.white),
                          ),
                          backgroundColor: Colors.red,
                          duration: Duration(seconds: 5),
                        ));
                        setState(() {
                          isGeneratingCode = false;
                        });
                      }
                    } catch (e) {
                      ScaffoldMessenger.of(context).showSnackBar(const SnackBar(
                        content: Text(
                          "Something went wrong!",
                          style: TextStyle(color: Colors.white),
                        ),
                        backgroundColor: Colors.red,
                        duration: Duration(seconds: 5),
                      ));
                    }
                  } else {
                    ScaffoldMessenger.of(context).showSnackBar(const SnackBar(
                      content: Text(
                        "Transaction cancelled!",
                        style: TextStyle(color: Colors.white),
                      ),
                      backgroundColor: Colors.red,
                      duration: Duration(seconds: 5),
                    ));
                  }
                })
          ]),
          onWebViewCreated: (WebViewController webViewController) async {
            _webViewController = webViewController;
            _loadhtmlFromString();
          },
        ),
      ),
    );
  }

  // _loadhtmlFromAssets() async {
  //   String filehtml = await rootBundle.loadString(filepath);
  //   _webViewController.loadUrl(Uri.dataFromString(filehtml,
  //           mimeType: 'text/html', encoding: Encoding.getByName('utf-8'))
  //       .toString());
  //   //flutter snackbars
  //   ScaffoldMessenger.of(context).showSnackBar(const SnackBar(
  //     content: Text("Loading..."),
  //     duration: Duration(seconds: 5),
  //   ));
  //   //delay for 5 seconds
  //   Future.delayed(const Duration(seconds: 5), () {
  //     loadfunction();
  //   });
  // }

  htmlString() {
    return '''
    <!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Ade Flutterwave</title>
  </head>
  <body>
    <h2 style="text-align:center;">Making payment, please wait</h2>
    <script src="https://checkout.flutterwave.com/v3.js"></script>
    <script>
      function test(params) {
        let d = document.querySelector("h2");
        d.innerHTML = 'Making payment for'+ params + ', <br>please wait...';
      }

      function sendBack(data) {
        messageHandler.postMessage(data);
      }

      function closeWebView(){
        //close the tab
        window.stop();
        var data = {
            "status" : "cancelled",
        };
        //stringify
        var dataString = JSON.stringify(data);
        messageHandler.postMessage(dataString);
      }

      function makePayment(tx_ref, amount, email, title, name, currency, icon, public_key, phone, payment_options) {
        test(title);

        FlutterwaveCheckout({
          public_key: public_key,
          tx_ref: tx_ref,
          amount: amount,
          currency: currency,
          payment_options: payment_options,
          // specified redirect URL
          //   redirect_url:
          //     "https://callbacks.piedpiper.com/flutterwave.aspx?ismobile=34",
          customer: {
            email: email,
            phone_number: phone,
            name: name
          },
          callback: function (data) {
            console.log(data);
            let dd = JSON.stringify(data);
            // console.log(dd);
             window.stop();
            sendBack(dd);
          },
          onclose: function () {
            // close modal
            closeWebView();
          },
          customizations: {
            title: title,
            description: "Payment for items in cart",
            logo: icon
          }
        });
      }
    </script>
  </body>
</html>
    ''';
  }

  _loadhtmlFromString() async {
    _webViewController.loadUrl(Uri.dataFromString(htmlString(),
            mimeType: 'text/html', encoding: Encoding.getByName('utf-8'))
        .toString());

    //flutter snackbars
    ScaffoldMessenger.of(context).showSnackBar(const SnackBar(
      content: Text("Loading..."),
      duration: Duration(seconds: 5),
    ));
    //delay for 5 seconds
    Future.delayed(const Duration(seconds: 5), () {
      loadfunction();
    });
  }
}


lib/example.dart
import 'package:flutter/material.dart';
import 'package:ade_flutterwave_working_version/core/ade_flutterwave.dart';

class AdeFlutterExample extends StatefulWidget {
  const AdeFlutterExample({Key? key}) : super(key: key);

  @override
  State<AdeFlutterExample> createState() => _AdeFlutterExampleState();
}

class _AdeFlutterExampleState extends State<AdeFlutterExample> {
  //controllers
  final TextEditingController _amountController = TextEditingController();
  final TextEditingController _emailController = TextEditingController();
  final TextEditingController _phoneController = TextEditingController();
  final TextEditingController _fullNameController = TextEditingController();

  @override
  Widget build(BuildContext context) {
    return Scaffold(
        appBar: AppBar(
          title: const Text('Ade Flutterwave payment'),
          backgroundColor: Colors.orange,
          foregroundColor: Colors.white,
        ),
        body: Padding(
          padding: const EdgeInsets.all(15.0),
          child: Form(
            child: Column(
              children: [
                //title
                const Text(
                  'Flutterwave payment',
                  style: TextStyle(
                    fontSize: 20,
                    fontWeight: FontWeight.bold,
                  ),
                ),
                TextFormField(
                  controller: _fullNameController,
                  decoration: const InputDecoration(
                    labelText: 'Name',
                  ),
                ),
                TextFormField(
                  controller: _emailController,
                  decoration: const InputDecoration(
                    labelText: 'Email',
                  ),
                ),
                TextFormField(
                  controller: _phoneController,
                  decoration: const InputDecoration(
                    labelText: 'Phone',
                  ),
                ),
                TextFormField(
                  controller: _amountController,
                  decoration: const InputDecoration(
                    labelText: 'Amount',
                  ),
                ),
                ElevatedButton(
                  onPressed: () {
                    //check if the fields are empty
                    if (_amountController.text.isEmpty ||
                        _emailController.text.isEmpty ||
                        _phoneController.text.isEmpty ||
                        _fullNameController.text.isEmpty) {
                      ScaffoldMessenger.of(context).showSnackBar(
                        const SnackBar(
                          content: Text('Please fill all the fields'),
                        ),
                      );
                      return;
                    }
                    //data
                    var data = {
                      'amount': _amountController.text,
                      'email': _emailController.text,
                      'phone': _phoneController.text,
                      'name': _fullNameController.text,
                      'payment_options': 'card, banktransfer, ussd',
                      'title': 'Flutterwave payment',
                      'currency': "NGN",
                      'tx_ref':
                          "AdeFlutterwave-${DateTime.now().millisecondsSinceEpoch}",
                      'icon':
                          "https://www.aqskill.com/wp-content/uploads/2020/05/logo-pde.png",
                      'public_key':
                          "FLWPUBK_TEST-e0787ab2e5b0b6fcb3d32ce465ad44d0-X",
                      'sk_key':
                          'FLWSECK_TEST-af1af523da3f141f894a26be4b071230-X'
                    };

                    Navigator.push(
                      context,
                      MaterialPageRoute(
                        builder: (context) => AdeFlutterWavePay(data),
                      ),
                    ).then((response) {
                      //response is the response from the payment
                      // print(response);
                    });
                  },
                  child: const Text('Pay'),
                ),
              ],
            ),
          ),
        ));
  }
}
